### ISSUE TITLE
AI: integrate open source facematching

### ISSUE THREAD

**Issue by urbien** [tradle/mycloud#22]
## Problem 
Face matching has become mainstream and open source is catching up with the proprietary products. 

1. To catch the price sensitive side of the market, we need to have an option of open source facematching. Such cases are self-KYC, KYC for individuals in developing world, CEX and DeFi.
2. EU has [now a law on AI](https://www.lawfareblog.com/artificial-intelligence-act-what-european-approach-ai). This law classified risk in 5 categories. Low risk category has no bearing on us, but others do:
    2.1. **Limited risk**. AI must be transparent. Person has the right to know, e.g. that they are talking to a bot, that their emotional state is being assessed.
   2.2. **High risk**. Biometric identification and categorization of natural persons. It looks like we will nee to get a certification mark (CE). Note that we might get an excluded from this requirement on the grounds of the purpose being crime detection and prevention (KYC), but we need to confirm that. But the law requires one to prove certain aspects of the AI system (below) which are easiest to prove with the open source implementations.

##  Burden of proof for high-risk AI to be compliant to EU law
> Data and Data Governance: High-risk AI systems must be developed using quality datasets, including those used for training, validating and testing the algorithm. Concretely, this quality requirement means that data must be relevant, representative, free of errors and complete. In addition, good data management practices such as paying particular attention to biases as well as to data gaps and data shortcomings are also mandatory. 
> 
> Transparency for Users: People who develop high-risk AI systems (“providers” under the proposed regulation) must disclose certain types of information to ensure proper use of AI systems. For example, providers must disclose information about the characteristics, capabilities and limitations of the AI system; the system’s intended purpose; and information necessary for its maintenance and care. 
> 
> Human Oversight: High-risk AI systems must be designed to be overseen by humans. Importantly, it does not mean that people must have a precise understanding of how AI systems—often described as black boxes—come to a particular decision. Instead, the focus is on an individual’s capacity to understand the main limitations of AI systems and ability to identify such shortcomings in a particular system. The overseeing duties include surveilling for automation bias problems, spotting anomalies or signs of dysfunctions, and deciding whether to override an AI system’s decision or to pull the “kill switch” if a system poses a threat to the safety and fundamental rights of people. 
> 
> Accuracy, Robustness and Cybersecurity: High-risk AI systems must achieve a level of accuracy, robustness and cybersecurity proportionate to their intended purpose. Providers of AI systems will be obligated to communicate accuracy metrics to those who use their AI systems. Backup or fail-safe plans to ensure sufficient robustness will also be required, as will technical solutions to prevent cybersecurity incidents such as data poisoning. 
> 
> Traceability and Auditability: Providers of high-risk AI systems must establish technical documentation containing information necessary to assess their compliance with the other requirements mentioned above. The extensive list of things that must be documented—like data management practices and risk management systems—can be found in Annex IV of the AI Act. Moreover, automatic recording of events (logs) is mandatory under the proposal.
> 
> Providers can comply with these requirements through a self-assessment procedure, except for remote biometric identification systems, which have to go through a more stringent process. Once the compliance assessment is done, the provider of an AI system completes an EU declaration of conformity and then can affix the CE marking of conformity and enter the European market. 
> 
> Compliance is not static, and the proposed regulation requires high-risk AI system providers to enact a postmarket monitoring process that actively evaluates the system’s compliance throughout its life cycle. 
n
## Implementation
look at @jacobgins's plugin that today uses HTTP to communicate with the facematching inference implementation (that is running under a docker in ECS on EC2)

**Comment by martinheidegger** [comment]
The Plugin in question with the EC2 instance is here: https://github.com/tradle/mycloud/blob/310058ad0b8ae212b56776ddbde860bfbfb44d15/src/in-house-bot/plugins/facial-recognition.ts#L261

**Comment by martinheidegger** [comment]
Note on the legal side: This is a start of a legal framework not a yet on track to become adopted by the member states.

<img width="600" alt="Screen Shot 2021-08-31 at 23 20 49" src="https://user-images.githubusercontent.com/914122/131519754-513580f0-6b61-4176-8c87-ea440e5f4753.png">

**Comment by urbien** [comment]
@martinheidegger right, needs to be adopted by the states in EU, but it is already adopted by the European Commission, and is a result of 3 years of legislative process.

**Comment by urbien** [comment]
@martinheidegger I need a status update on this project. What was completed, with references to the sources in repos, what still needs to be developed, what tests were developed, what issues are still not solved. I also need the list of all underlying frameworks and their versions that were used.

**Comment by martinheidegger** [comment]
Current state of development: Proof-of-Concept in branch of MyCloud running with a manually set-up docker-based lambda. Current result not working well. Status by sub-task:

- [ ] 1-to-1 face matching
    - [ ] Model training process
        - [x] Downloading training data
        - [x] Preparing training servers
        - [x] Initial training using the existing script. ← Two partial runs, once on AWS EC2 G5.2xlarge and G5.xlarge for performance comparison. partial model is currently stored in S3 bucket, can be used for inference, results not awesome.
        - [ ] Update of training script ← It turns out that big part (40%?) of the training process on the smallest AWS machine is the resizing/preparing of the input training data. This can be done by any machine really and can be applied in advance to the training data using spare computing resources. No need to run this on the expensive machine and expect this to be part of the required training process.
        - [ ] Finishing the training for the rest of the execution data.
        - [ ] Figure out how to get a commercial or OSS license for the training data @spwilko
    - [x] Inference Server
        - [x] [Python API] to execute Inference Model
        - [x] [Docker building](https://github.com/tradle/KYCDeepFace/blob/e85a55a06bfcc4086648998314f79ef65ddaa662/.github/workflows/deploy.yml#L18-L24) and deployment [on AWS](https://gallery.ecr.aws/h7s8u8m7/kycdeepface)
    - [ ] MyCloud Integration
       - [ ] JavaScript API to consume/use the KYCDeepface server: State Done, but not properly published yet.
       - [x] Configurable Face-Matching Plugin
       - [ ] Setting up Serverless to use Docker based Lambda configuration (Cloudformation)
    - [ ] [NIST Benchmarking](https://pages.nist.gov/frvt/html/frvt11.html) of the solution.
- [ ] 1-to-n face matching
    - [ ] Research serverless/lambda friendly ways to create/update 1-N matching (big research bar that I need to write down)
    - [ ] Review what to reuse existing partial implementation (that uses relational database)
    - [ ] Implementation 😅
         - [ ] Lookup logic
         - [ ] 1-N plugin for Mycloud
         - [ ] Documentation
- [ ] Use of ONNX models (Goal: use inference on mobile devices or in/with javascript lambdas)
    - [ ] Extract all image preprocessing steps into common logic
    - [ ] Publish new integation packages
- [ ] Further Research
    - [ ] Optimization of face pre-alignments: can better image-pre-processing help with this?
    - [ ] Evaluate differences to [InsightFace](https://github.com/deepinsight/insightface)

**Comment by urbien** [comment]
@martinheidegger regarding the problem of AWS Lambda limitation on how much data you can pass from the plugin in Lambda to a Lambda in a docker.

@mvayngrib suggested this:
Let's say you are working with the object PhotoID which has a large attachment.
- PhotoID._link is actually the key of this PhotoID json object in S3 objects bucket
- retrieve this json from S3 and one of its properties has a reference to S3 object with the image of the ID
This was you do not need to pass much between two Lambdas, and docker Lambda can retrieve all it needs in 2 s3 gets

There is an optimization that @mvayngrib will suggest to have only 1 S3 get but this can be done as a separate project.
